include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,160 @@
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_PACKET_POOL_HPP
+#define TORRENT_PACKET_POOL_HPP
+
+#include "libtorrent/config.hpp"
+
+#include "libtorrent/aux_/vector.hpp"
+
+namespace libtorrent
+{
+
+	// used for out-of-order incoming packets
+	// as well as sent packets that are waiting to be ACKed
+	struct packet
+	{
+		// the last time this packet was sent
+		time_point send_time;
+
+		// the number of bytes actually allocated in 'buf'
+		std::uint16_t allocated;
+
+		// the size of the buffer 'buf' points to
+		std::uint16_t size;
+
+		// this is the offset to the payload inside the buffer
+		// this is also used as a cursor to describe where the
+		// next payload that hasn't been consumed yet starts
+		std::uint16_t header_size;
+
+		// the number of times this packet has been sent
+		std::uint8_t num_transmissions:6;
+
+		// true if we need to send this packet again. All
+		// outstanding packets are marked as needing to be
+		// resent on timeouts
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+
+	struct packet_deleter
+	{
+		void operator()(packet* handle) const
+		{
+			std::free(handle);
+		}
+	};
+
+	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
+
+	template<std::size_t ALLOCATE_SIZE>
+	struct TORRENT_EXTRA_EXPORT packet_slab : private aux::vector<packet_ptr, std::size_t>
+	{
+		static const std::size_t allocate_size{ ALLOCATE_SIZE };
+
+		packet_slab(std::size_t limit) : m_limit(limit) { reserve(m_limit); }
+
+		void try_free(packet_ptr &p)
+		{
+			if (size() < m_limit)
+				push_back(std::move(p));
+		}
+
+		void try_alloc(packet_ptr &p)
+		{
+			if (empty())
+				p.reset(static_cast<packet*>(std::malloc(sizeof(packet) + allocate_size)));
+			else
+			{
+				p = std::move(back());
+				pop_back();
+			}
+		}
+	private:
+		std::size_t m_limit;
+	};
+
+	// single thread packet allocation packet pool
+	// can handle common cases of packet size by 3 pools
+	struct TORRENT_EXTRA_EXPORT packet_pool : private single_threaded
+	{
+		packet *acquire(std::size_t allocate)
+		{
+			TORRENT_ASSERT(is_single_thread());
+			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
+
+			packet_ptr p;
+
+			if (allocate <= m_syn_slab.allocate_size) { m_syn_slab.try_alloc(p); }
```

Comment:
doing that, you no longer need to assume that packets freed by this pool were allocated by it, and you can have == comparisons, and I think it becomes a bit more robust